### PR TITLE
feat: support LabelSelector type for NamespaceSelectors

### LINKED ISSUE #2481: Support complete label in namespaceSelectorwatch
*Description*:
`EnvoyGateway.Provider.Kubernetes.Wach.NamespaceSelectors`
NamespaceSelectors holds a list of labels that namespaces have to have in order to be watched.

The namespaceSelector can watch the namespace that has some key of label,  I think it might be better to use a map to achieve a complete label.

--

[optional *Relevant Links*:]
#1622

### PR DESCRIPTION
**Pull request by yeedove**
**What type of PR is this?**

**What this PR does / why we need it**:

In the current configuration, the `namespaceSelectors` field under `EnvoyGateway.Provider.Kubernetes.Wach` only supports label selection using the key, similar to the `Exists` operator in `LabelSelectorOperator`.

The `namespaceSelectors` field can support a more comprehensive label-based selection, allowing you to specify both `matchLabels` and `matchExpressions`
```yaml
namespaceSelectors:
  matchLabels:
    component: redis
  matchExpressions:
    - { key: tier, operator: In, values: [cache] }
    - { key: environment, operator: NotIn, values: [dev] }
```
[https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#LabelSelector](url)

**Which issue(s) this PR fixes**:

Fixes #2481

### CODE REVIEW

**Comment by ShyunnY** [comment]
LGTM

File: api/v1alpha1/envoygateway_types.go [hunk 1]
```diff
@@ -222,11 +222,10 @@ type KubernetesWatchMode struct {
...
 
-	// NamespaceSelectors holds a list of labels that namespaces have to have in order to be watched.
-	// Note this doesn't set the informer to watch the namespaces with the given labels. Informer still
-	// watches all namespaces. But the events for objects whose namespace do not match given labels
-	// will be filtered out. Precisely one of Namespaces and NamespaceSelectors must be set.
-	NamespaceSelectors []string `json:"namespaceSelectors,omitempty"`
+	// NamespaceSelectors holds the label selector used to dynamically select namespaces.
+	// Envoy Gateway will watch for namespaces matching the specified label selector.
+	// Precisely one of Namespaces and NamespaceSelectors must be set.
+	NamespaceSelectors *metav1.LabelSelector `json:"namespaceSelectors,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/envoygateway_types.go:228]: should this field be renamed to `namespaceSelector` ?

Comment (yeedove) [reply api/v1alpha1/envoygateway_types.go]: yes, done

**Comment by zirain** [comment]
EG use squash merge model, please kindly don't force push everytime for better review.

**Comment by yeedove** [comment]
> EG use squash merge model, please kindly don't force push everytime for better review.

I will pay attention to this.  Can you help me look at the `coverage-test`? thank

**Comment by zirain** [comment]
```console
--- FAIL: TestNamespaceSelectorProvider (67.82s)
    kubernetes_test.go:1372: 
        	Error Trace:	/home/runner/work/gateway/gateway/internal/provider/kubernetes/kubernetes_test.go:1372
        	Error:      	Condition never satisfied
        	Test:       	TestNamespaceSelectorProvider
2024-01-25T11:14:20.404Z	INFO	kubernetes/predicates.go:39	gatewayclass has matching controller name, processing	{"name": "test-gc"}
2024-01-25T11:14:20.404Z	INFO	kubernetes/predicates.go:43	bypassing reconciliation due to controller name	{"controller": "not.configured/controller"}
2024-01-25T11:14:20.406Z	DPANIC	kubernetes/predicates.go:123	non-string key argument passed to logging, ignoring all later arguments	{"invalid key": "not.configured/controller"}
2024-01-25T11:14:20.406Z	INFO	kubernetes/predicates.go:123	gatewayclass controller name
2024-01-25T11:14:20.406Z	INFO	kubernetes/predicates.go:124	gatewayclass name for gateway doesn't match configured name	{"namespace": "", "name": "scheduled-status-test"}
2024-01-25T11:14:20.407Z	DPANIC	kubernetes/predicates.go:123	non-string key argument passed to logging, ignoring all later arguments	{"invalid key": "not.configured/controller"}
2024-01-25T11:14:20.407Z	INFO	kubernetes/predicates.go:123	gatewayclass controller name
2024-01-25T11:14:20.407Z	INFO	kubernetes/predicates.go:124	gatewayclass name for gateway doesn't match configured name	{"namespace": "", "name": "scheduled-status-test"}
2024-01-25T11:14:20.417Z	INFO	kubernetes/routes.go:266	processing HTTPRoute	{"namespace": "test", "name": "test"}
2024-01-25T11:14:20.418Z	INFO	kubernetes/routes.go:266	processing HTTPRoute	{"namespace": "test", "name": "test"}
2024-01-25T11:14:20.418Z	INFO	kubernetes/routes.go:266	processing HTTPRoute	{"namespace": "test", "name": "test"}
2024-01-25T11:14:20.419Z	INFO	kubernetes/routes.go:137	processing GRPCRoute	{"namespace": "test", "name": "test"}
```

**Review by shawnh2 — COMMENTED** [review]
thanks for adding this, generally lgtm, just some nits.

**Comment by yeedove** [comment]
@shawnh2  done, thank!

**Review by shawnh2 — APPROVED** [review]
LGTM, thanks

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**